Small step towards VR sunflares

JonnyOThan opened the pull request:
-use the correct eye to calculate the sun's viewport position
-don't reset extinction until the right eye is done
-however something isn't working properly; the sun's viewport position doesn't seem to get updated in the shader properties for the right eye.  Using Shader.SetGlobalVector instead of Material.SetVector seems to work, but that would break multi-star systems.

JonnyOThan committed:
Small step toward sunflares and VR compatibility
-use the correct eye to calculate the sun's viewport position
-don't reset extinction until the right eye is done
-however something isn't working properly; the sun's viewport position doesn't seem to get updated in the shader properties for the right eye.  Using Shader.SetGlobalVector instead of Material.SetVector seems to work, but that would break multi-star systems.

JonnyOThan commented:
Hey, is the sunflare shader in the OldShaders folder still representative of what scatterer is using?  Cause I'm wondering if the sunViewPortPos is no longer explicitly declared by the shader as a property.

LGhassen commented:
Hey, I haven't checked this PR yet, I will check it the next time I'm modding.

For the flare shader yes it is still exactly the same. The sunViewPortPos and everything under CGInclude are variables to be passed from CPU code. I don't know what you mean exactly about property? If you mean declaring them in the properties section at the top of the shader, I never used that and don't know what it's for.

JonnyOThan commented:
Yeah that's exactly what I mean - the properties are the things passed in from the CPU side.  For whatever reason, trying to change this one in between rendering the left & right eye doesn't seem to work.

It's super strange; seems sort of like a bug in unity...Interestingly the stock unity lens flares work perfectly in VR.  Do you think it's reasonable to use those to handle the rendering?  I think you could still drive all the parameters that scatterer needs for extinction etc.  For now I'm just adding a cfg patch to KerbalVR to disable scatterer's sunflares; maybe I can switch to using FlareReplacer instead, but I haven't looked into it too deeply.

LGhassen commented:
It could be that this function accounts only for one eye? https://github.com/LGhassen/Scatterer/blob/dev/scatterer/Effects/SunFlare/SunFlare.cs#L136 have you tried debugging if the value changes on plugin side before being changed as a parameter?

JonnyOThan commented:
Yes, you'll notice in the PR that I changed that to use the current camera's current active eye.  I added logging and verified with the debugger that different values are being sent to the shader for each eye, but the shader only ever sees the values for the left eye (verified with RenderDoc).

LGhassen commented:
This is very weird. Is there any difference in behaviour between tracking station/map view and flight? Does map view work while flight is broken?

LGhassen commented:
Otherwise try to make 2 materials one for each eye and switch them?

LGhassen commented:
Since we established that the issue is material related (like with the clouds), I'll just pass the local or scaled position of the sun to the shader and let the shader calculate directly the viewport position with it's current MVP matrix, that should fix it.

LGhassen commented:
@JonnyOThan check my latest commit
